pkgs/test_core/lib/src/runner/vm/test_compiler.dart

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:test_api/backend.dart'; // ignore: deprecated_member_use
+import 'package:frontend_server_client/frontend_server_client.dart';
+
+import '../package_version.dart';
+import '../../util/package_config.dart';
+
+/// A request to the [TestCompiler] for recompilation.
+class _CompilationRequest {
+  _CompilationRequest(this.mainUri, this.result, this.metadata);
+
+  Uri mainUri;
+  Metadata metadata;
+  Completer<CompilationResponse> result;
+}
+
+class CompilationResponse {
+  final String? compilerOutput;
+  final int errorCount;
+  final Uri? kernelOutputUri;
+
+  CompilationResponse(
+      {this.compilerOutput, this.errorCount = 0, this.kernelOutputUri});
+}
+
+class TestCompiler {
+  TestCompiler(this._dillCachePath)
+      : _outputDillDirectory =
+            Directory.systemTemp.createTempSync('dart_test.') {
+    _outputDill = File(p.join(_outputDillDirectory.path, 'output.dill'));
+    _compilerController.stream.listen(_onCompilationRequest);
```

Comment:
I don't follow what the stream is accomplishing for us when we also have a separate queue.

Can we either omit the stream

```dart
Future<CompilationResponse> compile(Uri mainDart, Metadata metadata) {   
    if (_isClosed) {
      return Future.value(null);
    }
    final completer = Completer<CompilationResponse>();
    if(_compilationQueue.isEmpty) Future(_drainQueue);
    _compilationQueue.add(_CompilationRequest(mainDart, completer, metadata));
    return completer.future;
 }
 
 Future<void> _drainQueue() async {
   while (_compilationQueue.isNotEmpty) {
     // ...
 ```
  
  Or use the stream as the queue
  
  ```dart
  void _runCompiles(Stream<_CompilationRequest> requests) async {
    await for (final request in requests) {
      var firstCompile = false;
      CompileResult? compilerOutput;
      final contents =
          await _generateEntrypoint(request.mainUri, request.metadata);
      final tempFile = File(p.join(_outputDillDirectory.path, 'test.dart'))
        ..writeAsStringSync(contents);
      // Rest of _onCompilationRequest implementation...
 ```

Or we can use `package:pool` to handle this in `compile` and skip both the stream and the queue

```dart
  final _compilePool = Pool(1);
  Future<CompilationResponse> compile(Uri mainDart, Metadata metadata) {
    return _compilePool.withResource(() => _compile(mainDart, metadata));
  }
  Future<CompileResult> _compile(Uri mainDart, Metadata metadata) async {
    var firstCompile = false;
    CompileResult? compilerOutput;
    final contents = await _generateEntrypoint(mainUri, metadata);
    final tempFile = File(p.join(_outputDillDirectory.path, 'test.dart'))
      ..writeAsStringSync(contents);

    if (_frontendServerClient == null) {
      compilerOutput = await _createCompiler(tempFile.uri);
      firstCompile = true;
    } else {
      compilerOutput =
          await _frontendServerClient!.compile(<Uri>[tempFile.uri]);
    }
    // The client is guaranteed initialized at this point.
    final client = _frontendServerClient!;
    final outputPath = compilerOutput?.dillOutput;
    if (outputPath == null) {
      return CompilationResponse(
          compilerOutput: compilerOutput?.compilerOutputLines.join('\n'),
          errorCount: compilerOutput?.errorCount ?? 0);
    } else {
      final outputFile = File(outputPath);
      final kernelReadyToRun = await outputFile.copy('${tempFile.path}.dill');
      final testCache = File(_dillCachePath);
      // Keep the cache file up-to-date and use the size of the kernel file
      // as an approximation for how many packages are included. Larger files
      // are prefered, since re-using more packages will reduce the number of
      // files the frontend server needs to load and parse.
      if (firstCompile ||
          !testCache.existsSync() ||
          (testCache.lengthSync() < outputFile.lengthSync())) {
        if (!testCache.parent.existsSync()) {
          testCache.parent.createSync(recursive: true);
        }
        await outputFile.copy(_dillCachePath);
      }
      client.accept();
      client.reset();
      return CompilationResponse(
          compilerOutput: compilerOutput?.compilerOutputLines.join('\n'),
          errorCount: compilerOutput?.errorCount ?? 0,
          kernelOutputUri: kernelReadyToRun.absolute.uri);
    }
  }
```

Reply:
Updated to use package:pool